src/sparsebase/io/mtx_writer.cc

```diff
@@ -0,0 +1,390 @@
+void MTXWriter<IDType, NNZType, ValueType>::WriteCOO(
+    format::COO<IDType, NNZType, ValueType> *coo) const {
+      
+      //illegal parameter checks
+      if (object_ != "matrix" && object_ != "vector") 
+      {
+        throw utils::WriterException("Illegal value for the 'object' option in matrix market header");
+      }
+      else if (object_ == "vector")
+      {
+        throw utils::WriterException("Matrix market writer does not currently support writing vectors.");
+      }
+      if (format_ != "array" && format_ != "coordinate") 
+      {
+        throw utils::WriterException("Illegal value for the 'format' option in matrix market header");
+      }
+      if (field_ != "real" && field_ != "double" && field_ != "complex" && field_ != "integer" && field_ != "pattern") 
+      {
+        throw utils::WriterException("Illegal value for the 'field' option in matrix market header");
+      }
+      if (symmetry_ != "general" && symmetry_ != "symmetric" && symmetry_ != "skew-symmetric" && symmetry_ != "hermitian") 
+      {
+        throw utils::WriterException("Illegal value for the 'symmetry' option in matrix market header");
+      }
+      if (format_ == "array" && field_ == "pattern") 
+      {
+        throw utils::WriterException("Matrix market files with array format cannot have the field ""'pattern' ");
+      }
+      if (format_ == "array" && symmetry_ != "general") 
+      {
+        throw utils::WriterException("Matrix market files with array format cannot have the property 'symmetry' ");
+      }
+      if (symmetry_ == "hermitian") 
+      {
+        throw utils::WriterException("Matrix market writer does not currently support hermitian symmetry.");
+      }
+
+      std::ofstream mtxFile;
+      mtxFile.open(filename_);
+      
+      //header
+      mtxFile << "%%MatrixMarket " << object_ << " " << format_ << " " << field_ << " " << symmetry_ << "\n";
+      
+      //TODO add warning when given field and actual data format is not same (integer vs float etc)
+
+      if constexpr (std::is_same_v<ValueType, void>)
+      {
+        if (field_ == "pattern")
+          throw utils::WriterException("Cannot write an MTX with void ValueType, unless field is pattern.");
+      }
+
+      auto dimensions = coo->get_dimensions();
+      IDType* row = coo->get_row();
+      IDType* col = coo->get_col();
+      ValueType* val = coo->get_vals();
+
+      if (val != NULL && field_ == "pattern")
+      {
+        utils::Logger logger(typeid(this));
+        logger.Log("Pattern selected but values given.", utils::LOG_LVL_WARNING);
+      }
+
+      //Symmetry check
+      bool saidSymmetric = (symmetry_ == "symmetric" || symmetry_ == "skew-symmetric" || symmetry_ == "hermitian");
+      int NNZ = coo->get_num_nnz();
+      int count_symmetric = 0;
+      int count_diagonal = 0;
+      int is_diagonal_all_zero = true;
+      if (saidSymmetric && dimensions[0] == dimensions[1]) 
+      {
```

Comment:
What if `saidSymmetric == True` but `dimensions[0]!=dimensions[1]`. I think in that case, an exception should be thrown as well.